File: python/multi_channel_sync.py
#!/usr/bin/env python3
"""
Multi-Channel Todo Synchronization
Coordinates todos across WhatsApp, Email, Google Sheets, and GitHub Issues
"""

import os
import json
import subprocess
from datetime import datetime
from env_loader import load_env

# Import our integration modules
from whatsapp_todo_integration import WhatsAppTodoManager
from email_todo_integration import EmailTodoManager
from google_sheets_integration import GoogleSheetsTodoManager

class MultiChannelTodoSync:
    def __init__(self):
        load_env()
        self.sync_log_file = "sync_log.json"
        
        # Initialize all channel managers
        self.whatsapp_manager = WhatsAppTodoManager()
        self.email_manager = EmailTodoManager()
        self.sheets_manager = GoogleSheetsTodoManager()
        
        self.load_sync_log()
    
    def load_sync_log(self):
        """Load synchronization log"""
        try:
            if os.path.exists(self.sync_log_file):
                with open(self.sync_log_file, 'r') as f:
                    self.sync_log = json.load(f)
            else:
                self.sync_log = {
                    'last_sync': None,
                    'sync_history': []
                }
        except Exception as e:
            print(f"Error loading sync log: {e}")
            self.sync_log = {'last_sync': None, 'sync_history': []}
    
    def save_sync_log(self):
        """Save synchronization log"""
        try:
            with open(self.sync_log_file, 'w') as f:
                json.dump(self.sync_log, f, indent=2)
        except Exception as e:
            print(f"Error saving sync log: {e}")
    
    def sync_all_channels(self):
        """Synchronize todos across all channels"""
        print("🔄 Starting multi-channel synchronization...")
        sync_timestamp = datetime.now().isoformat()
        
        sync_results = {
            'timestamp': sync_timestamp,
            'whatsapp_processed': 0,
            'emails_processed': 0,
            'sheets_synced': 0,
            'github_created': 0,
            'errors': []
        }
        
        try:
            # 1. Process new WhatsApp messages (if webhook data available)
            print("📱 Checking WhatsApp todos...")
            whatsapp_todos = self.whatsapp_manager.list_todos("919742814697")  # Default number
            sync_results['whatsapp_processed'] = len(whatsapp_todos)
            
            # 2. Process new emails
            print("📧 Processing emails...")
            email_success = self.email_manager.process_emails()
            if email_success:
                email_todos = self.email_manager.list_todos()
                sync_results['emails_processed'] = len(email_todos)
            else:
                sync_results['errors'].append("Email processing failed")
            
            # 3. Sync with Google Sheets
            print("📊 Syncing Google Sheets...")
            try:
                # Add WhatsApp todos to sheets
                for todo in whatsapp_todos[-5:]:  # Last 5 todos
                    if todo['status'] == 'pending':
                        self.sheets_manager.add_todo(
                            f"[WhatsApp] {todo['description']}",
                            notes=f"From: {todo['phone_number']}"
                        )
                        sync_results['sheets_synced'] += 1
                
                # Add email todos to sheets
                email_todos = self.email_manager.list_todos()
                for todo in email_todos[-5:]:  # Last 5 todos
                    if todo['status'] == 'pending':
                        self.sheets_manager.add_todo(
                            f"[Email] {todo['description']}",
                            notes=f"From: {todo['sender_email']}"
                        )
                        sync_results['sheets_synced'] += 1
                        
            except Exception as e:
                sync_results['errors'].append(f"Sheets sync error: {e}")
            
            # 4. Create GitHub issues for high-priority todos
            print("🐙 Creating GitHub issues...")
            try:
                self.create_github_issues_from_channels()
                sync_results['github_created'] = 1  # Simplified count
            except Exception as e:
                sync_results['errors'].append(f"GitHub sync error: {e}")
            
            # 5. Generate unified todo report
            print("📋 Generating unified report...")
            self.generate_unified_report()
            
            # Update sync log
            self.sync_log['last_sync'] = sync_timestamp
            self.sync_log['sync_history'].append(sync_results)
            
            # Keep only last 10 sync records
            if len(self.sync_log['sync_history']) > 10:
                self.sync_log['sync_history'] = self.sync_log['sync_history'][-10:]
            
            self.save_sync_log()
            
            # Print summary
            print(f"\n✅ Multi-channel sync completed!")
            print(f"📊 Summary:")
            print(f"   • WhatsApp todos: {sync_results['whatsapp_processed']}")
            print(f"   • Email todos: {sync_results['emails_processed']}")
            print(f"   • Sheets synced: {sync_results['sheets_synced']}")
            print(f"   • GitHub issues: {sync_results['github_created']}")
            
            if sync_results['errors']:
                print(f"⚠️  Errors encountered: {len(sync_results['errors'])}")
                for error in sync_results['errors']:
                    print(f"     • {error}")
            
            return True
            
        except Exception as e:
            print(f"❌ Sync failed: {e}")
            sync_results['errors'].append(f"General sync error: {e}")
            return False
    
    def create_github_issues_from_channels(self):
        """Create GitHub issues from high-priority channel todos"""
        try:
            # Get high-priority todos from all channels
            priority_todos = []
            
            # WhatsApp todos (recent ones)
            whatsapp_todos = self.whatsapp_manager.list_todos("919742814697")
            for todo in whatsapp_todos[-3:]:  # Last 3
                if todo['status'] == 'pending':
                    priority_todos.append({
                        'title': f"[WhatsApp] {todo['description']}",
                        'body': f"Created via WhatsApp from {todo['phone_number']}\nCreated: {todo['created_at']}",
                        'source': 'whatsapp'
                    })
            
            # Email todos (recent ones)
            email_todos = self.email_manager.list_todos()
            for todo in email_todos[-3:]:  # Last 3
                if todo['status'] == 'pending':
                    priority_todos.append({
                        'title': f"[Email] {todo['description']}",
                        'body': f"From: {todo['sender_email']}\nSubject: {todo.get('email_subject', 'N/A')}\nCreated: {todo['created_at']}",
                        'source': 'email'
                    })
            
            # Create GitHub issues
            for todo in priority_todos:
                try:
                    cmd = [
                        'gh', 'issue', 'create',
                        '--repo', 'hebbarp/todo-management',
                        '--title', todo['title'],
                        '--body', todo['body'],
                        '--label', f"source:{todo['source']},status:pending"
                    ]
                    subprocess.run(cmd, check=True, capture_output=True)
                    print(f"✅ Created GitHub issue: {todo['title']}")
                except subprocess.CalledProcessError:
                    print(f"⚠️ Could not create GitHub issue (gh CLI may not be available)")
                    break
                    
        except Exception as e:
            print(f"Error creating GitHub issues: {e}")
    
    def generate_unified_report(self):
        """Generate a unified report across all channels"""
        report_file = f"unified_todo_report_{datetime.now().strftime('%Y%m%d_%H%M%S')}.json"
        
        try:
            # Collect todos from all sources
            whatsapp_todos = self.whatsapp_manager.list_todos("919742814697")
            email_todos = self.email_manager.list_todos()
            sheets_todos = self.sheets_manager.list_todos()
            
            unified_report = {
                'generated_at': datetime.now().isoformat(),
                'summary': {
                    'total_todos': len(whatsapp_todos) + len(email_todos) + len(sheets_todos),
                    'whatsapp_count': len(whatsapp_todos),
                    'email_count': len(email_todos),
                    'sheets_count': len(sheets_todos)
                },
                'todos': {
                    'whatsapp': whatsapp_todos[-10:],  # Last 10
                    'email': email_todos[-10:],        # Last 10
                    'sheets': sheets_todos[-10:] if sheets_todos else []  # Last 10
                },
                'sync_status': self.sync_log
            }
            
            # Save report
            with open(report_file, 'w') as f:
                json.dump(unified_report, f, indent=2)
            
            print(f"📊 Unified report saved: {report_file}")
            
            # Generate human-readable summary
            summary_text = f"""📋 UNIFIED TODO REPORT
================================
Generated: {datetime.now().strftime("%Y-%m-%d %H:%M:%S")}

📊 Summary:
• Total todos across all channels: {unified_report['summary']['total_todos']}
• WhatsApp todos: {unified_report['summary']['whatsapp_count']}
• Email todos: {unified_report['summary']['email_count']}
• Google Sheets todos: {unified_report['summary']['sheets_count']}

🔄 Last sync: {self.sync_log.get('last_sync', 'Never')}

📱 Recent WhatsApp Todos:"""
            
            for todo in whatsapp_todos[-5:]:
                status_emoji = "✅" if todo['status'] == 'completed' else "⏳"
                summary_text += f"\n   {status_emoji} #{todo['id']}: {todo['description']}"
            
            summary_text += f"\n\n📧 Recent Email Todos:"
            for todo in email_todos[-5:]:
                status_emoji = "✅" if todo['status'] == 'completed' else "⏳"
                summary_text += f"\n   {status_emoji} #{todo['id']}: {todo['description']}"
            
            print(f"\n{summary_text}")
            
            # Save human-readable summary
            summary_file = f"unified_summary_{datetime.now().strftime('%Y%m%d_%H%M%S')}.txt"
            with open(summary_file, 'w') as f:
                f.write(summary_text)
            
            return unified_report
            
        except Exception as e:
            print(f"Error generating unified report: {e}")
            return None
    
    def send_daily_digest(self, recipient_email):
        """Send daily digest email with all channel updates"""
        try:
            report = self.generate_unified_report()
            if not report:
                return False
            
            # Create digest email
            digest_subject = f"Daily Todo Digest - {datetime.now().strftime('%B %d, %Y')}"
            digest_body = f"""📊 Your Daily Todo Digest

🔄 Multi-Channel Summary:
• Total active todos: {report['summary']['total_todos']}
• WhatsApp: {report['summary']['whatsapp_count']} todos
• Email: {report['summary']['email_count']} todos  
• Google Sheets: {report['summary']['sheets_count']} todos

📱 Recent WhatsApp Activity:"""
            
            whatsapp_todos = report['todos']['whatsapp']
            for todo in whatsapp_todos[-3:]:
                digest_body += f"\n   • {todo['description']}"
            
            digest_body += f"\n\n📧 Recent Email Activity:"
            email_todos = report['todos']['email']
            for todo in email_todos[-3:]:
                digest_body += f"\n   • {todo['description']}"
            
            digest_body += f"\n\n🔄 Last sync: {self.sync_log.get('last_sync', 'Never')}"
            digest_body += f"\n\n💡 Reply to this email to add new todos to your list!"
            
            # Send via email manager
            success = self.email_manager.send_status_email(
                recipient_email, 
                digest_subject, 
                digest_body
            )
            
            return success
            
        except Exception as e:
            print(f"Error sending daily digest: {e}")
            return False
    
    def emergency_backup(self):
        """Create emergency backup of all todos"""
        backup_timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
        backup_file = f"emergency_backup_{backup_timestamp}.json"
        
        try:
            backup_data = {
                'backup_created': datetime.now().isoformat(),
                'whatsapp_todos': self.whatsapp_manager.todos,
                'email_todos': self.email_manager.todos,
                'sheets_data': self.sheets_manager.list_todos(),
                'sync_log': self.sync_log
            }
            
            with open(backup_file, 'w') as f:
                json.dump(backup_data, f, indent=2)
            
            print(f"💾 Emergency backup created: {backup_file}")
            return backup_file
            
        except Exception as e:
            print(f"❌ Backup failed: {e}")
            return None

def main():
    """Main synchronization function"""
    print("🚀 Multi-Channel Todo Sync Starting...")
    print("=" * 50)
    
    sync_manager = MultiChannelTodoSync()
    
    # Perform full synchronization
    success = sync_manager.sync_all_channels()
    
    if success:
        print("\n🎉 Multi-channel synchronization completed!")
        
        # Create backup
        backup_file = sync_manager.emergency_backup()
        if backup_file:
            print(f"💾 Backup created: {backup_file}")
        
        # Optionally send digest email
        gmail_user = os.getenv('GMAIL_USER')
        if gmail_user:
            print(f"📧 Sending digest to {gmail_user}...")
            sync_manager.send_daily_digest(gmail_user)
        
        return True
    else:
        print("❌ Synchronization failed")
        return False

if __name__ == "__main__":
    main()

File: todo-system-php/includes/MultiChannelSync.php
<?php
require_once 'config.php';
require_once 'TodoManager.php';
require_once 'WhatsAppIntegration.php';
require_once 'EmailIntegration.php';
require_once 'GoogleSheetsIntegration.php';

/**
 * Multi-Channel Todo Synchronization
 * Coordinates todos across WhatsApp, Email, Google Sheets, and GitHub Issues
 */
class MultiChannelSync {
    private $todoManager;
    private $whatsAppIntegration;
    private $emailIntegration;
    private $sheetsIntegration;
    private $syncLogFile;
    
    public function __construct() {
        $this->todoManager = new TodoManager();
        $this->whatsAppIntegration = new WhatsAppIntegration();
        $this->emailIntegration = new EmailIntegration();
        $this->sheetsIntegration = new GoogleSheetsIntegration();
        $this->syncLogFile = Config::get('data_dir') . 'sync_log.json';
    }
    
    /**
     * Perform complete multi-channel synchronization
     */
    public function syncAllChannels() {
        Config::log("Starting multi-channel synchronization");
        
        $syncTimestamp = date('c');
        $syncResults = [
            'timestamp' => $syncTimestamp,
            'whatsapp_processed' => 0,
            'emails_processed' => 0,
            'sheets_synced' => 0,
            'github_created' => 0,
            'errors' => []
        ];
        
        try {
            // 1. Process emails for new todos
            Config::log("Processing emails...");
            $emailSuccess = $this->emailIntegration->processEmails();
            if ($emailSuccess) {
                $emailTodos = $this->todoManager->getTodos(null, 'email', 10);
                $syncResults['emails_processed'] = count($emailTodos);
            } else {
                $syncResults['errors'][] = "Email processing failed";
            }
            
            // 2. Sync with Google Sheets
            Config::log("Syncing Google Sheets...");
            try {
                $sheetsSynced = $this->sheetsIntegration->syncFromOtherSources(10);
                $syncResults['sheets_synced'] = $sheetsSynced;
            } catch (Exception $e) {
                $syncResults['errors'][] = "Sheets sync error: " . $e->getMessage();
            }
            
            // 3. Create GitHub issues for high-priority todos
            Config::log("Creating GitHub issues...");
            try {
                $githubCreated = $this->createGitHubIssuesFromChannels();
                $syncResults['github_created'] = $githubCreated;
            } catch (Exception $e) {
                $syncResults['errors'][] = "GitHub sync error: " . $e->getMessage();
            }
            
            // 4. Generate unified report
            Config::log("Generating unified report...");
            $reportFile = $this->generateUnifiedReport();
            
            // 5. Create emergency backup
            $backupFile = $this->createEmergencyBackup();
            
            // Update sync log
            $this->updateSyncLog($syncResults);
            
            // Print summary
            $this->printSyncSummary($syncResults, $reportFile, $backupFile);
            
            Config::log("Multi-channel sync completed successfully");
            return true;
            
        } catch (Exception $e) {
            Config::log("Sync failed: " . $e->getMessage(), 'ERROR');
            $syncResults['errors'][] = "General sync error: " . $e->getMessage();
            $this->updateSyncLog($syncResults);
            return false;
        }
    }
    
    /**
     * Create GitHub issues from high-priority channel todos
     */
    private function createGitHubIssuesFromChannels() {
        $priorityTodos = [];
        $createdCount = 0;
        
        // Get recent todos from all channels
        $recentTodos = $this->todoManager->getTodos('pending', null, 5);
        
        foreach ($recentTodos as $todo) {
            if ($todo['source'] !== 'manual') {
                $priorityTodos[] = [
                    'title' => "[{$todo['source']}] {$todo['description']}",
                    'body' => $this->formatGitHubIssueBody($todo),
                    'source' => $todo['source']
                ];
            }
        }
        
        // Create GitHub issues
        foreach ($priorityTodos as $todo) {
            try {
                $command = sprintf(
                    'gh issue create --repo hebbarp/todo-management --title %s --body %s --label %s 2>/dev/null',
                    escapeshellarg($todo['title']),
                    escapeshellarg($todo['body']),
                    escapeshellarg("source:{$todo['source']},status:pending")
                );
                
                exec($command, $output, $returnCode);
                
                if ($returnCode === 0) {
                    Config::log("Created GitHub issue: {$todo['title']}");
                    $createdCount++;
                } else {
                    Config::log("Could not create GitHub issue (gh CLI may not be available)", 'DEBUG');
                    break; // Don't try to create more if gh CLI is not available
                }
                
            } catch (Exception $e) {
                Config::log("Error creating GitHub issue: " . $e->getMessage(), 'DEBUG');
                break;
            }
        }
        
        return $createdCount;
    }
    
    /**
     * Format GitHub issue body
     */
    private function formatGitHubIssueBody($todo) {
        $body = "Created via {$todo['source']} integration\n";
        $body .= "Created: {$todo['created_at']}\n";
        $body .= "Priority: {$todo['priority']}\n\n";
        
        if (!empty($todo['source_data'])) {
            $sourceData = $todo['source_data'];
            
            if (isset($sourceData['phone_number'])) {
                $body .= "WhatsApp: {$sourceData['phone_number']}\n";
            }
            
            if (isset($sourceData['sender_email'])) {
                $body .= "Email: {$sourceData['sender_email']}\n";
            }
            
            if (isset($sourceData['email_subject'])) {
                $body .= "Subject: {$sourceData['email_subject']}\n";
            }
            
            if (isset($sourceData['notes'])) {
                $body .= "Notes: {$sourceData['notes']}\n";
            }
        }
        
        return $body;
    }
    
    /**
     * Generate unified report across all channels
     */
    private function generateUnifiedReport() {
        $reportTimestamp = date('Y-m-d_H-i-s');
        $reportFile = Config::get('backups_dir') . "unified_todo_report_$reportTimestamp.json";
        
        try {
            // Collect data from all sources
            $allTodos = $this->todoManager->getTodos();
            $whatsappTodos = $this->todoManager->getTodos(null, 'whatsapp');
            $emailTodos = $this->todoManager->getTodos(null, 'email');
            $sheetsTodos = $this->sheetsIntegration->getTodos();
            $stats = $this->todoManager->getStats();
            
            $unifiedReport = [
                'generated_at' => date('c'),
                'summary' => [
                    'total_todos' => count($allTodos),
                    'whatsapp_count' => count($whatsappTodos),
                    'email_count' => count($emailTodos),
                    'sheets_count' => count($sheetsTodos),
                    'pending_count' => $stats['pending'],
                    'completed_count' => $stats['completed'],
                    'completion_rate' => $stats['completion_rate']
                ],
                'todos' => [
                    'whatsapp' => array_slice($whatsappTodos, -10), // Last 10
                    'email' => array_slice($emailTodos, -10),       // Last 10
                    'sheets' => array_slice($sheetsTodos, -10)      // Last 10
                ],
                'sync_status' => $this->getSyncLog()
            ];
            
            // Save report
            file_put_contents($reportFile, json_encode($unifiedReport, JSON_PRETTY_PRINT));
            
            // Generate human-readable summary
            $this->generateHumanReadableSummary($unifiedReport, $reportTimestamp);
            
            Config::log("Unified report saved: $reportFile");
            return $reportFile;
            
        } catch (Exception $e) {
            Config::log("Error generating unified report: " . $e->getMessage(), 'ERROR');
            return null;
        }
    }
    
    /**
     * Generate human-readable summary
     */
    private function generateHumanReadableSummary($report, $timestamp) {
        $summaryFile = Config::get('backups_dir') . "unified_summary_$timestamp.txt";
        
        $summary = "📋 UNIFIED TODO REPORT\n";
        $summary .= "================================\n";
        $summary .= "Generated: " . date('Y-m-d H:i:s') . "\n\n";
        
        $summary .= "📊 Summary:\n";
        $summary .= "• Total todos across all channels: {$report['summary']['total_todos']}\n";
        $summary .= "• WhatsApp todos: {$report['summary']['whatsapp_count']}\n";
        $summary .= "• Email todos: {$report['summary']['email_count']}\n";
        $summary .= "• Google Sheets todos: {$report['summary']['sheets_count']}\n";
        $summary .= "• Pending: {$report['summary']['pending_count']}\n";
        $summary .= "• Completed: {$report['summary']['completed_count']}\n";
        $summary .= "• Completion Rate: {$report['summary']['completion_rate']}%\n\n";
        
        $syncLog = $this->getSyncLog();
        $lastSync = $syncLog['last_sync'] ?? 'Never';
        $summary .= "🔄 Last sync: $lastSync\n\n";
        
        $summary .= "📱 Recent WhatsApp Todos:\n";
        foreach ($report['todos']['whatsapp'] as $todo) {
            $statusEmoji = $todo['status'] === 'completed' ? '✅' : '⏳';
            $summary .= "   $statusEmoji #{$todo['id']}: {$todo['description']}\n";
        }
        
        $summary .= "\n📧 Recent Email Todos:\n";
        foreach ($report['todos']['email'] as $todo) {
            $statusEmoji = $todo['status'] === 'completed' ? '✅' : '⏳';
            $summary .= "   $statusEmoji #{$todo['id']}: {$todo['description']}\n";
        }
        
        file_put_contents($summaryFile, $summary);
        echo $summary;
    }
    
    /**
     * Create emergency backup of all todos
     */
    private function createEmergencyBackup() {
        $backupTimestamp = date('Y-m-d_H-i-s');
        $backupFile = Config::get('backups_dir') . "emergency_backup_$backupTimestamp.json";
        
        try {
            $backupData = [
                'backup_created' => date('c'),
                'database_todos' => $this->todoManager->getTodos(),
                'database_stats' => $this->todoManager->getStats(),
                'sheets_data' => $this->sheetsIntegration->getTodos(),
                'sheets_stats' => $this->sheetsIntegration->getStats(),
                'sync_log' => $this->getSyncLog()
            ];
            
            file_put_contents($backupFile, json_encode($backupData, JSON_PRETTY_PRINT));
            
            Config::log("Emergency backup created: $backupFile");
            return $backupFile;
            
        } catch (Exception $e) {
            Config::log("Backup failed: " . $e->getMessage(), 'ERROR');
            return null;
        }
    }
    
    /**
     * Update sync log
     */
    private function updateSyncLog($syncResults) {
        try {
            $syncLog = $this->getSyncLog();
            
            $syncLog['last_sync'] = $syncResults['timestamp'];
            
            // Add to history
            if (!isset($syncLog['sync_history'])) {
                $syncLog['sync_history'] = [];
            }
            
            $syncLog['sync_history'][] = $syncResults;
            
            // Keep only last 10 sync records
            if (count($syncLog['sync_history']) > 10) {
                $syncLog['sync_history'] = array_slice($syncLog['sync_history'], -10);
            }
            
            file_put_contents($this->syncLogFile, json_encode($syncLog, JSON_PRETTY_PRINT));
            
        } catch (Exception $e) {
            Config::log("Error updating sync log: " . $e->getMessage(), 'ERROR');
        }
    }
    
    /**
     * Get sync log
     */
    private function getSyncLog() {
        if (file_exists($this->syncLogFile)) {
            $content = file_get_contents($this->syncLogFile);
            return json_decode($content, true) ?? ['last_sync' => null, 'sync_history' => []];
        }
        
        return ['last_sync' => null, 'sync_history' => []];
    }
    
    /**
     * Print sync summary
     */
    private function printSyncSummary($syncResults, $reportFile, $backupFile) {
        echo "\n✅ Multi-channel sync completed!\n";
        echo "📊 Summary:\n";
        echo "   • Email todos: {$syncResults['emails_processed']}\n";
        echo "   • Sheets synced: {$syncResults['sheets_synced']}\n";
        echo "   • GitHub issues: {$syncResults['github_created']}\n";
        
        if (!empty($syncResults['errors'])) {
            echo "⚠️  Errors encountered: " . count($syncResults['errors']) . "\n";
            foreach ($syncResults['errors'] as $error) {
                echo "     • $error\n";
            }
        }
        
        echo "\n📊 Generated files:\n";
        if ($reportFile) {
            echo "   • Unified report: " . basename($reportFile) . "\n";
        }
        if ($backupFile) {
            echo "   • Emergency backup: " . basename($backupFile) . "\n";
        }
        
        echo "\n🎉 Sync process complete!\n\n";
    }
    
    /**
     * Send daily digest email
     */
    public function sendDailyDigest($recipientEmail) {
        try {
            $stats = $this->todoManager->getStats();
            $pendingTodos = $this->todoManager->getTodos('pending', null, 5);
            $completedToday = $this->getTodosCompletedToday();
            
            $digestSubject = "Daily Todo Digest - " . date('F d, Y');
            $digestBody = "📊 Your Daily Todo Digest\n\n";
            $digestBody .= "🔄 Multi-Channel Summary:\n";
            $digestBody .= "• Total active todos: {$stats['total']}\n";
            $digestBody .= "• Pending: {$stats['pending']}\n";
            $digestBody .= "• Completed: {$stats['completed']}\n";
            $digestBody .= "• Completion rate: {$stats['completion_rate']}%\n\n";
            
            $digestBody .= "📋 Recent Pending Tasks:\n";
            foreach ($pendingTodos as $todo) {
                $digestBody .= "   • {$todo['description']} (from {$todo['source']})\n";
            }
            
            $digestBody .= "\n✅ Completed Today (" . count($completedToday) . "):\n";
            foreach ($completedToday as $todo) {
                $digestBody .= "   • {$todo['description']}\n";
            }
            
            if (empty($completedToday)) {
                $digestBody .= "   • No todos completed today\n";
            }
            
            $syncLog = $this->getSyncLog();
            $lastSync = $syncLog['last_sync'] ?? 'Never';
            $digestBody .= "\n🔄 Last sync: $lastSync\n";
            $digestBody .= "\n💡 Reply to this email to add new todos to your list!";
            
            $success = $this->emailIntegration->sendStatusEmail(
                $recipientEmail,
                $digestSubject,
                $digestBody
            );
            
            return $success;
            
        } catch (Exception $e) {
            Config::log("Error sending daily digest: " . $e->getMessage(), 'ERROR');
            return false;
        }
    }
    
    /**
     * Get todos completed today
     */
    private function getTodosCompletedToday() {
        try {
            $pdo = Config::getDatabaseConnection();
            $sql = "SELECT * FROM todos WHERE status = 'completed' AND DATE(completed_at) = DATE('now') ORDER BY completed_at DESC";
            $stmt = $pdo->prepare($sql);
            $stmt->execute();
            
            $todos = $stmt->fetchAll(PDO::FETCH_ASSOC);
            
            foreach ($todos as &$todo) {
                $todo['source_data'] = json_decode($todo['source_data'], true);
            }
            
            return $todos;
        } catch (Exception $e) {
            Config::log("Error fetching today's completed todos: " . $e->getMessage(), 'ERROR');
            return [];
        }
    }
    
    /**
     * Cleanup old data
     */
    public function cleanup($daysOld = 30) {
        $cleanupResults = [
            'todos_deleted' => 0,
            'emails_deleted' => 0,
            'logs_rotated' => false
        ];
        
        try {
            // Cleanup old completed todos
            $cleanupResults['todos_deleted'] = $this->todoManager->cleanup($daysOld);
            
            // Cleanup old processed emails
            $pdo = Config::getDatabaseConnection();
            $sql = "DELETE FROM processed_emails WHERE processed_at < DATE('now', '-$daysOld days')";
            $stmt = $pdo->prepare($sql);
            $stmt->execute();
            $cleanupResults['emails_deleted'] = $stmt->rowCount();
            
            // Rotate log files if they're too large
            $logFile = Config::get('logs_dir') . 'system.log';
            if (file_exists($logFile) && filesize($logFile) > 10 * 1024 * 1024) { // 10MB
                $archiveFile = Config::get('logs_dir') . 'system_' . date('Y-m-d') . '.log';
                rename($logFile, $archiveFile);
                $cleanupResults['logs_rotated'] = true;
            }
            
            Config::log("Cleanup completed: " . json_encode($cleanupResults));
            return $cleanupResults;
            
        } catch (Exception $e) {
            Config::log("Error during cleanup: " . $e->getMessage(), 'ERROR');
            return $cleanupResults;
        }
    }
}
?>